Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
-    [priv->allWorkDone signal];
-
     return self;
 }
 
-- (NSOperationQueue*)_doMainWork {
-    memset(priv->myQueues, 0, sizeof(priv->myQueues));
+- (void)dealloc {
+    [_name release];
 
-    BOOL didWork;
+    for (NSOperation* op in _operations.get()) {
+        [op removeObserver:self forKeyPath:@"isFinished" context:nil];
+        [op removeObserver:self forKeyPath:@"isReady" context:nil];
+    }
 
-    id innerPool = [[NSAutoreleasePool alloc] init];
+    if (_underlyingQueue) {
+        dispatch_release(_underlyingQueue);
+    }
 
-    do {
-        didWork = FALSE;
+    [super dealloc];
+}
 
-        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-            [priv->suspendedCondition lock];
-            while (priv->isSuspended) {
-                [priv->suspendedCondition wait];
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        NSOperation* op = static_cast<NSOperation*>(object);
+        if (op.isFinished) {
+            // Minor optimization:
+            // Because of constraints in addOperation:, it is safe to assume that there is exactly one instance of op in the array
+            // removeObject: keeps going after finding first instance of object, which is unnecessary work
+            // Use indexOfObject:op instead
+            @synchronized(_operations.get()) {
+                NSUInteger index = [_operations indexOfObject:op];
+                if (index == NSNotFound) {
+                    TraceError(g_logTag, L"Observed an operation finishing that was not in this operation queue");
+                    return;
+                }
+
+                [self willChangeValueForKey:@"operations"];
+                [self willChangeValueForKey:@"operationCount"];
+                [_operations removeObjectAtIndex:index];
+                [self didChangeValueForKey:@"operationCount"];
+                [self didChangeValueForKey:@"operations"];
             }
-            [priv->suspendedCondition unlock];
 
-            if (RunOperationFromLists(&priv->myQueues[i], &priv->queues[i], &priv->curOperation)) {
-                didWork = TRUE;
-            }
+            // Clean up state from addOperation:
+            [op removeObserver:self forKeyPath:@"isFinished" context:nil];
+            [op removeObserver:self forKeyPath:@"isReady" context:nil];
         }
-    } while (didWork);
-
-    [innerPool release];
-
-    return self;
+    } else if ([keyPath isEqualToString:@"isReady"]) {
+        NSOperation* op = static_cast<NSOperation*>(object);
+        [self _startOperation:op];
+    }
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)hasMoreWork {
-    if (priv->curOperation != nil) {
-        return TRUE;
++ (NSOperationQueue*)currentQueue {
+    NSThread* currentThread = [NSThread currentThread];
+    if ([currentThread isMainThread]) {
+        return [[self class] mainQueue];
```

Comment:
While it won't break anything, using `[self class]` in a `+` method is ambiguous as to whether it should be a meta-metaclass, or just a metaclass.